### PR TITLE
Allow custom particles for the plants

### PR DESCRIPTION
**Pull request by nbrichau**

### CODE REVIEW

File: src/main/java/com/infinityraider/agricraft/api/v1/plant/IAgriPlant.java [hunk 1]
```diff
@@ -258,6 +259,11 @@ void getClipProducts(@Nonnull Consumer<ItemStack> products, @Nonnull ItemStack c
     @Nonnull
     ResourceLocation getSeedModel();
 
+    /**
+     * @return The particle effect for the plant
+     */
+    List<AgriParticleEffect> getParticleEffects();
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/agricraft/api/v1/plant/IAgriPlant.java:265]: No referencing to external classes in the API please.
Only Vanilla, Forge, and API classes.

File: src/main/java/com/infinityraider/agricraft/content/core/BlockCropPlant.java [hunk 2]
```diff
@@ -173,4 +179,34 @@ public ActionResultType onBlockActivated(BlockState state, World world, BlockPos
...
+        if (!optional.isPresent()) {
+            return;
+        }
+        IAgriCrop crop = optional.get();
+        if (crop.hasPlant() && crop.isMature()) {
+            List<AgriParticleEffect> particleEffects = crop.getPlant().getParticleEffects();
+            for (AgriParticleEffect particleEffect : particleEffects) {
+                double deltaX = (rand.nextBoolean() ? 1 : -1) * particleEffect.getDeltaX() * rand.nextDouble();
+                double deltaY = (rand.nextBoolean() ? 1 : -1) * particleEffect.getDeltaY() * rand.nextDouble();
+                double deltaZ = (rand.nextBoolean() ? 1 : -1) * particleEffect.getDeltaZ() * rand.nextDouble();
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/agricraft/content/core/BlockCropPlant.java:195]: Put these in the the final if, as it is not always reached, in which cases these operations are done for nothing.

File: src/main/java/com/infinityraider/agricraft/impl/v1/plant/NoPlant.java [hunk 3]
```diff
@@ -193,6 +194,11 @@ public ResourceLocation getSeedModel() {
         return this.seed_model;
     }
 
+    @Override
+    public List<AgriParticleEffect> getParticleEffects() {
+        return new ArrayList<>();
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/agricraft/impl/v1/plant/NoPlant.java:199]: Use Collections.emptyList(), or ImmutableList.of(), no need to make a new object to which people could be adding stuff.

**Review by InfinityRaider — CHANGES_REQUESTED** [review]
This is going to need modifications, instead of using a getter in IAgriPlant, make a method spawnparticles() instead and pass it the IAgriCrop object.
Call this method from the block class, and love the current implementation to JsonPlant. The implementation in NoPlant can be empty

**Comment by nbrichau** [comment]
Hope the changes are good to you.
I used an IAgriStage and not an IAgriCropPlant in IAgriCore#spawnParticles because receiving a stage is enough.

File: src/main/java/com/infinityraider/agricraft/api/v1/plant/IAgriPlant.java [hunk 4]
```diff
@@ -260,9 +261,14 @@ void getClipProducts(@Nonnull Consumer<ItemStack> products, @Nonnull ItemStack c
...
-     * @return The particle effect for the plant
+     * Spawn custom particles if the current growth stage allow it.
+     *
+     * @param world the curent world of the plant.
+     * @param pos the current position of the plant.
+     * @param rand a random for use in rng.
+     * @param stage the current growth stage of the plant.
      */
-    List<AgriParticleEffect> getParticleEffects();
+    void spawnParticles(World world, BlockPos pos, Random rand, IAgriGrowthStage stage);
```
Comment (InfinityRaider) [inline src/main/java/com/infinityraider/agricraft/api/v1/plant/IAgriPlant.java:271]: Just pass the IAgriCrop instance (and maybe the Random object), the crop instance gives access to all these parameters and more as well.

**Comment by InfinityRaider** [comment]
> 
> 
> Hope the changes are good to you.
> I used an IAgriStage and not an IAgriCropPlant in IAgriCore#spawnParticles because receiving a stage is enough.

The point of passing in the crop instance is to give people who implement plants via the API all the liberty they might require.
There is no reason to code in limitations.

**Comment by nbrichau** [comment]
Ok, I understand.

**Review by InfinityRaider — APPROVED** [review]

**Pull request merged**